{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Obtaining and Processing Network Reconstructions from KBase\n",
    "\n",
    "KBase (http://kbase.us) is a powerful resource for obtaining genome-scale network reconstructions for microbial genomes. These reconstructions are distributed as SBML files, which must be processed prior to reverse ecology analysis. This notebook describes how to obtain reconstructions from KBase, and how to process them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtaining and Preparing SBML Files\n",
    "\n",
    "Briefly, genomes (as fasta files) for your organisms of interest can be pushed from your computer to KBase. Once there, a KBase Narrative (iPython notebook) can be used to build reconstructions for your genomes. To do so, follow the instructions in `Perl/README.md`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtaining and Preparing SBML Files: Update January 2016\n",
    "\n",
    "To validate predictions made by this reverse ecology pipeline, I propose to map metatranscriptomic reads from OMD-TOIL to KBase-annotated genomes. Retrieving the genome annotations proved to be quite a pain!\n",
    "\n",
    "* The KBase Narrative mentioned above ([click here](https://narrative.kbase.us/narrative/ws.9599.obj.2)) no longer contains annotated versions of the genomes.\n",
    "\n",
    "* Furthermore, the code cells in that narrative generate errors during the annotation, model-building, and downloading steps.\n",
    "\n",
    "I have attempted to communicate with the KBase team to resolve some of these issues, but have received no response.\n",
    "\n",
    "Thus, I created new KBase Narratives for manual annotation, model-building, and downloading. For each notebook (linked below), I copied the genomes (contigs) from the [old narrative](https://narrative.kbase.us/narrative/ws.9599.obj.2) and manually ran the \"Annotate Contigs\" and \"Build Metabolic Model\" KBase apps. I then downloaded the annotated genomes (Genbank format) and models (SBML and tsv formats).\n",
    "\n",
    "* SAGs: [KBase Narrative](https://narrative.kbase.us/narrative/ws.12259.obj.1), workspace ID: joshamilton:1452727482251\n",
    "\n",
    "* Mendota MAGs: [KBase Narrative](https://narrative.kbase.us/narrative/ws.12268.obj.1), workspace ID: joshamilton:1452793144933\n",
    "\n",
    "* Trout Bog MAGs [KBase Narrative](https://narrative.kbase.us/narrative/ws.12270.obj.1), workspace ID: joshamilton:1452798604003\n",
    "\n",
    "* MAGs from other research groups [KBase Narrative](https://narrative.kbase.us/narrative/ws.12271.obj.1), workspace ID: joshamilton:1452801292037\n",
    "\n",
    "Once the genomes were downloaded, I further converted the Genbank-formatted genomes to fasta nucleotide (ffn), fasta amino acid (ffa), and gff format. The scripts to do so are in the refGenomes/scripts folder."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing SBML Files\n",
    "\n",
    "Reconstructions from KBase require further processing before they are suitable for use in reverse ecology. The function below does a nunmber of things:\n",
    "1. Reformat gene locus tags\n",
    "2. Remove biomass, exchange, transport, spontaneous, DNA/RNA biosynthesis reactions and their corresponding genes\n",
    "3. Import metabolite formulas\n",
    "4. Check mass- and charge-balancing of reactions in the reconstruction\n",
    "5. Remove trailing 0s from reaction and metabolite names\n",
    "\n",
    "The post-processing has a major shortcoming. When KBase detects that one or more subunits of a complex are present, it creates a \"full\" GPR by adding 'Unknown' genes for the other subunits. CobraPy currently lacks functions to remove the genes. As such, these model should not be used to perform any simulations which rely on GPRs.\n",
    "\n",
    "As output, the code returns processed SBML files in the 'processedDataDir' folder. Also returns a summary of the model sizes, in the 'summaryStatsDir' folder.\n",
    "\n",
    "The first chunk of code identifies imports the Python packages necessary for this analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Import special features for iPython\n",
    "import sys\n",
    "sys.path.append('../Python')\n",
    "import matplotlib\n",
    "\n",
    "# Import Python modules \n",
    "# These custom-written modules should have been included with the package\n",
    "# distribution. \n",
    "import sbmlFunctions as sf\n",
    "import metadataFunctions as mf\n",
    "\n",
    "# Define local folder structure for data input and processing.\n",
    "processedDataDir = 'ProcessedModelFiles'\n",
    "rawModelDir = 'RawModelFiles'\n",
    "summaryStatsDir = 'DataSummaries'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we call a function which processes each SBML file and preps it for analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing model AAA023D18, 1 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model AAA023J06, 2 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model AAA024D14, 3 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model AAA027E14, 4 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model AAA027J17, 5 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model AAA027L06, 6 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model AAA027L17, 7 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model AAA027M14, 8 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model AAA028A23, 9 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model AAA028E20, 10 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model AAA028G02, 11 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model AAA028I14, 12 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model AAA028K15, 13 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model AAA028P02, 14 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "All reactions are balanced\n",
      "Processing model AAA041L13, 15 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model AAA044D11, 16 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model AAA044N04, 17 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model AAA044O16, 18 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model AAA278I18, 19 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model AAA278O22, 20 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model AB141P03, 21 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model acIB-AMD-6, 22 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model acIB-AMD-7, 23 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model acMicro-1, 24 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "All reactions are balanced\n",
      "Processing model acMicro-4, 25 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model BIN_10, 26 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model BIN_15, 27 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model bin_29_2_acIII, 28 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "All reactions are balanced\n",
      "Processing model bin_7_acI-B2, 29 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint10152, 30 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "All reactions are balanced\n",
      "Processing model MEint1091, 31 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint11576, 32 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint1363, 33 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint14260, 34 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint1719, 35 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint1953, 36 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "All reactions are balanced\n",
      "Processing model MEint2297, 37 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model MEint2538, 38 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "All reactions are balanced\n",
      "Processing model MEint2729, 39 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint283, 40 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model MEint3119, 41 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint3163, 42 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint3864, 43 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint4252, 44 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model MEint885, 45 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model Rhodoluna, 46 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "All reactions are balanced\n",
      "Processing model TBepi149, 47 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model TBepi2057, 48 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model TBepi2754, 49 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model TBepi2973, 50 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model TBepi3207, 51 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Re-balancing on the basis of protons\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model TBepi3256, 52 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model TBepi3475, 53 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model TBepi4163, 54 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model TBepi4208, 55 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model TBhypo2152, 56 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3180, 57 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3219, 58 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3463, 59 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3765, 60 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "Manually correcting an imbalance\n",
      "Re-balancing on the basis of protons\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3838, 61 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model TBhypo680, 62 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n",
      "Processing model TBhypo9906, 63 of 63\n",
      "The remaining extracellular metabolites are:\n",
      "cpd00067_e0\n",
      "All reactions are balanced\n"
     ]
    }
   ],
   "source": [
    "sf.processSBMLforRE(rawModelDir, processedDataDir, summaryStatsDir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Pruning Currency Metabolites\n",
    "\n",
    "The next step is to prune currency metabolites from the metabolic network, in order for the network's directed graph to better reflect physiological metabolic transformations. We adopt an approach outlined by Ma et al, and later adopted by Borenstein et al in their \"reverse ecology\" paper.\n",
    "\n",
    "__Provide some illustrative examples of why this is necessary.__\n",
    "\n",
    "#### References\n",
    "1. Borenstein, E., Kupiec, M., Feldman, M. W., & Ruppin, E. (2008). Large-scale reconstruction and phylogenetic analysis of metabolic environments. Proceedings of the National Academy of Sciences, 105(38), 14482–14487. http://doi.org/10.1073/pnas.0806162105\n",
    "2. Ma, H., Zeng, A. P., & Network, A. A.-. (2003). Reconstruction of metabolic networks from genome data and analysis of their global structure for various organisms. Bioinformatics, 19(2), 270–277. http://doi.org/10.1093/bioinformatics/19.2.270\n",
    "3. Ma, H. W., & Zeng, A.-P. (2003). The connectivity structure, giant strong component and centrality of metabolic networks. Bioinformatics, 19(11), 1423–1430. http://doi.org/10.1093/bioinformatics/btg177\n",
    "\n",
    "Reaction pruning relies on checking each reaction against a set of criteria, and updating the stoichiometry as appropriate. Some criteria can easily be automated, such as removal of the following currency metabolites:\n",
    "\n",
    "| Metabolite | KBase Compound ID | \n",
    "|---|---|\n",
    "| H2O | | \n",
    "| CO2 | |\n",
    "| H2CO3 | |\n",
    "| O2 | |\n",
    "| H2O2 | |\n",
    "| NH3 | |\n",
    "| Nitrite | |\n",
    "| Nitrate | |\n",
    "| Nitric oxide | |\n",
    "| H2S | |\n",
    "| Sulfite | |\n",
    "| Sulfate | |\n",
    "| H+ | |\n",
    "| H2 | |\n",
    "| Phosphate | |\n",
    "| Pyrophosphate | |\n",
    "\n",
    "The script below loops over the set of ProcessedModelFiles and removes these metabolites from their associated reactions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Other pruning steps cannot be done programatically. Consider the criteria below:\n",
    "\n",
    "| Metabolite | Criteria | KBase Compound ID | \n",
    "|---|---|---|\n",
    "| CoA | Except reactions in CoA synthesis pathway |\n",
    "| NTP/NDP | As carrier for phosphate group transfer |\n",
    "| NTP/NMP | As carrier for phosphate group transfer |\n",
    "| NAD(P)/NAD(P)H | As carrier for hydrogen transfer |\n",
    "| FAD/FADH | As carrier for hydrogen transfer |\n",
    "| Ferrocytochrome/Ferricytochrome | As carrier for hydrogen transfer |\n",
    "| Reduced ferredoxin/Oxidized ferredoxin | As carrier for hydrogen transfer |\n",
    "| Glutathione/Oxidized glutathione | As carrier for hydrogen transfer |\n",
    "| Dihydrobiopterin/Tetrahydrobiopterin | As carrier for hydrogen transfer |\n",
    "| Glutamate/Oxoglutarate | As carrier for amino group transfer |\n",
    "| Glutamine/Glutamate | As carrier for amino group transfer |\n",
    "| Pyruvate/Alanine | As carrier for amino group transfer |\n",
    "| 5,10-Methenyl-THF, 5,10-Methylene-THF, THF |  Except reactions in CoA synthesis pathway |\n",
    "| 10-Formyl-THF, 5-Formyl-THF, THF |  Except reactions in CoA synthesis pathway |\n",
    "| 5-Formimino-THF, 5-Methyl-THF, THF |  Except reactions in CoA synthesis pathway |\n",
    "| S-Adenosyl-L-methionine, S-Adenosyl-L-homocysteine |  Except reactions in CoA synthesis pathway |\n",
    "| Adenosine 3',5'-bisphosphate/3'-Phosphoadenylyl sulfate | As carrier for sulfate group transfer |\n",
    "__Note__: This list may not be exhaustive, as the KBase reaction database may contain other cofactors and carriers not identifed by Ma et al in the KEGG database.\n",
    "\n",
    "Here, each reaction must be individually inspected to see if it meets the criteria. The script below identifies reactions containing the currency metabolites above, allowing the user to assess each reaction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
