{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Obtaining and Processing Network Reconstructions from KBase\n",
    "\n",
    "KBase (http://kbase.us) is a powerful resource for obtaining genome-scale network reconstructions for microbial genomes. These reconstructions are distributed as SBML files, which must be processed prior to reverse ecology analysis. This notebook describes how to obtain reconstructions from KBase, and how to process them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtaining and Preparing SBML Files\n",
    "\n",
    "Briefly, genomes (as fasta files) for your organisms of interest can be pushed from your computer to KBase. Once there, a KBase Narrative (iPython notebook) can be used to build reconstructions for your genomes. To do so, follow the instructions in `Perl/README.md`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtaining and Preparing SBML Files: Update January 2016\n",
    "\n",
    "To validate predictions made by this reverse ecology pipeline, I propose to map metatranscriptomic reads from OMD-TOIL to KBase-annotated genomes. Retrieving the genome annotations proved to be quite a pain!\n",
    "\n",
    "* The KBase Narrative mentioned above ([click here](https://narrative.kbase.us/narrative/ws.9599.obj.2)) no longer contains annotated versions of the genomes.\n",
    "\n",
    "* Furthermore, the code cells in that narrative generate errors during the annotation, model-building, and downloading steps.\n",
    "\n",
    "I have attempted to communicate with the KBase team to resolve some of these issues, but have received no response.\n",
    "\n",
    "Thus, I created new KBase Narratives for manual annotation, model-building, and downloading. For each notebook (linked below), I copied the genomes (contigs) from the [old narrative](https://narrative.kbase.us/narrative/ws.9599.obj.2) and manually ran the \"Annotate Contigs\" and \"Build Metabolic Model\" KBase apps. I then downloaded the annotated genomes (Genbank format) and models (SBML and tsv formats).\n",
    "\n",
    "* SAGs: [KBase Narrative](https://narrative.kbase.us/narrative/ws.12259.obj.1), workspace ID: joshamilton:1452727482251\n",
    "\n",
    "* Mendota MAGs: [KBase Narrative](https://narrative.kbase.us/narrative/ws.12268.obj.1), workspace ID: joshamilton:1452793144933\n",
    "\n",
    "* Trout Bog MAGs [KBase Narrative](https://narrative.kbase.us/narrative/ws.12270.obj.1), workspace ID: joshamilton:1452798604003\n",
    "\n",
    "* MAGs from other research groups [KBase Narrative](https://narrative.kbase.us/narrative/ws.12271.obj.1), workspace ID: joshamilton:1452801292037\n",
    "\n",
    "Once the genomes were downloaded, I further converted the Genbank-formatted genomes to fasta nucleotide (ffn), fasta amino acid (ffa), and gff format. The scripts to do so are in the refGenomes/scripts folder."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing SBML Files\n",
    "\n",
    "Reconstructions from KBase require further processing before they are suitable for use in reverse ecology. The function below does a nunmber of things:\n",
    "1. Reformat gene locus tags\n",
    "2. Remove biomass, exchange, spontaneous, DNA/RNA biosynthesis reactions and their corresponding genes\n",
    "3. Import metabolite formulas\n",
    "4. Check mass- and charge-balancing of reactions in the reconstruction\n",
    "5. Remove trailing 0s from reaction and metabolite names\n",
    "\n",
    "The post-processing has a major shortcoming. When KBase detects that one or more subunits of a complex are present, it creates a \"full\" GPR by adding 'Unknown' genes for the other subunits. CobraPy currently lacks functions to remove the genes. As such, these model should not be used to perform any simulations which rely on GPRs.\n",
    "\n",
    "As output, the code returns processed SBML files in the 'processedDataDir' folder. Also returns a summary of the model sizes, in the 'summaryStatsDir' folder.\n",
    "\n",
    "The first chunk of code identifies imports the Python packages necessary for this analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Import special features for iPython\n",
    "import sys\n",
    "sys.path.append('../Python')\n",
    "import matplotlib\n",
    "%matplotlib inline\n",
    "\n",
    "# Import Python modules \n",
    "# These custom-written modules should have been included with the package\n",
    "# distribution. \n",
    "import sbmlFunctions as sf\n",
    "import metadataFunctions as mf\n",
    "\n",
    "# Define local folder structure for data input and processing.\n",
    "processedDataDir = 'ProcessedModelFiles'\n",
    "rawModelDir = 'RawModelFiles'\n",
    "summaryStatsDir = 'DataSummaries'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we call a function which processes each SBML file and preps it for analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing model AAA023D18, 1 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA023J06, 2 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA024D14, 3 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA027D23, 4 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA027E14, 5 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA027F04, 6 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA027J17, 7 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA027L06, 8 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA027L17, 9 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA027M14, 10 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA028A23, 11 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA028C09, 12 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA028E20, 13 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA028G02, 14 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA028I14, 15 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA028K15, 16 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA028N15, 17 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA028P02, 18 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA041L13, 19 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA044D11, 20 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA044N04, 21 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA044O16, 22 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA278I18, 23 of 72\n",
      "All reactions are balanced\n",
      "Processing model AAA278O22, 24 of 72\n",
      "All reactions are balanced\n",
      "Processing model AB141P03, 25 of 72\n",
      "All reactions are balanced\n",
      "Processing model acAcidi, 26 of 72\n",
      "All reactions are balanced\n",
      "Processing model acAMD-2, 27 of 72\n",
      "All reactions are balanced\n",
      "Processing model acAMD-5, 28 of 72\n",
      "All reactions are balanced\n",
      "Processing model acIB-AMD-6, 29 of 72\n",
      "All reactions are balanced\n",
      "Processing model acIB-AMD-7, 30 of 72\n",
      "All reactions are balanced\n",
      "Processing model acMicro-1, 31 of 72\n",
      "All reactions are balanced\n",
      "Processing model acMicro-4, 32 of 72\n",
      "All reactions are balanced\n",
      "Processing model BIN_10, 33 of 72\n",
      "All reactions are balanced\n",
      "Processing model BIN_15, 34 of 72\n",
      "All reactions are balanced\n",
      "Processing model bin_29_2_acIII, 35 of 72\n",
      "All reactions are balanced\n",
      "Processing model bin_7_acI-B2, 36 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint10152, 37 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint1091, 38 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint11576, 39 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint1363, 40 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint14260, 41 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint1719, 42 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint1953, 43 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint2297, 44 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint2538, 45 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint2729, 46 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint283, 47 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint3119, 48 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint3163, 49 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint3864, 50 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint4252, 51 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint6647, 52 of 72\n",
      "All reactions are balanced\n",
      "Processing model MEint885, 53 of 72\n",
      "All reactions are balanced\n",
      "Processing model Rhodoluna, 54 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBepi149, 55 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBepi2057, 56 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBepi2754, 57 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBepi2973, 58 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBepi3207, 59 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBepi3256, 60 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBepi3475, 61 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBepi4163, 62 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBepi4208, 63 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBhypo2152, 64 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3180, 65 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3219, 66 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3463, 67 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3765, 68 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3815, 69 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBhypo3838, 70 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBhypo680, 71 of 72\n",
      "All reactions are balanced\n",
      "Processing model TBhypo9906, 72 of 72\n",
      "All reactions are balanced\n"
     ]
    }
   ],
   "source": [
    "sf.processSBMLforRE(rawModelDir, processedDataDir, summaryStatsDir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
