{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merging of Metabolic Network Graphs\n",
    "\n",
    "The purpose of this notebook is to identify criteria for classifying GFMs (without 16S sequence data) as members of a particular tribe, based on shared amino acid identity (ANI). The metabolic models of these genomes will then be merged to give a model which is better representative of a particular tribe. Our reverse ecology analysis will then be performed on these merged models."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging Genomes Based on ANI\n",
    "Pairwise ANI were computed by Sarah using the method in Goris et al IJSEM 2007. This notebook examines pairwise ANI values between all members within a tribe to identify a minimum ANI criterion for adding a new genome to the tribe. Then I identify genomes which are candidates for inclusion within a tribe (based on phylogeny) and determine whether or not they can be included within a tribe (based on ANI).\n",
    "\n",
    "#### References\n",
    "1. Konstantinidis, K. T., & Tiedje, J. M. (2005). Genomic insights that advance the species definition for prokaryotes. Proceedings of the National Academy of Sciences, 102(7), 2567–2572.\n",
    "2. Goris, J., Konstantinidis, K. T., Klappenbach, J. A., Coenye, T., Vandamme, P., & Tiedje, J. M. (2007). DNA-DNA hybridization values and their relationship to whole-genome sequence similarities. International Journal of Systematic and Evolutionary Microbiology, 57(1), 81–91.\n",
    "\n",
    "#### TO DO\n",
    "1. Extend this analysis to examine coverage as well as ANI.\n",
    "\n",
    "The first chunk of code identifies imports the Python packages necessary for this analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Import special features for iPython\n",
    "import sys\n",
    "sys.path.append('../Python')\n",
    "import matplotlib\n",
    "%matplotlib inline\n",
    "\n",
    "# Import Python modules \n",
    "# These custom-written modules should have been included with the package\n",
    "# distribution. \n",
    "import pairwiseANIFunctions as ANI\n",
    "\n",
    "# Define local folder structure for data input and processing.\n",
    "externalDataDir = 'ExternalData'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using SAGs to Compute Cutoffs\n",
    "\n",
    "I want to identify an ANI cutoff, such that any two genomes belonging to the same tribe have an ANI greater than this value. The function below computes the minimum and maximum pairwise ANI for all genomes belonging to the same tribe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>Num Samples</th>\n",
       "      <th>Max ANI</th>\n",
       "      <th>Min ANI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Iluma-A2</th>\n",
       "      <td>[AAA027E14]</td>\n",
       "      <td>1</td>\n",
       "      <td>99.97</td>\n",
       "      <td>99.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Iluma-B1</th>\n",
       "      <td>[AAA027L17]</td>\n",
       "      <td>1</td>\n",
       "      <td>99.54</td>\n",
       "      <td>99.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Luna1-A2</th>\n",
       "      <td>[AAA028P02]</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-A1</th>\n",
       "      <td>[AAA027M14, AAA278O22]</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>79.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-A5</th>\n",
       "      <td>[AAA028G02, AAA044O16]</td>\n",
       "      <td>2</td>\n",
       "      <td>99.95</td>\n",
       "      <td>86.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-A6</th>\n",
       "      <td>[AAA028E20, AAA028I14]</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>81.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-A7</th>\n",
       "      <td>[AAA023J06, AAA024D14, AAA041L13, AAA044N04]</td>\n",
       "      <td>4</td>\n",
       "      <td>100</td>\n",
       "      <td>87.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-B1</th>\n",
       "      <td>[AAA023D18, AAA027J17, AAA027L06, AAA028A23, A...</td>\n",
       "      <td>6</td>\n",
       "      <td>100</td>\n",
       "      <td>79.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-B4</th>\n",
       "      <td>[AAA044D11]</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Samples Num Samples  \\\n",
       "Iluma-A2                                        [AAA027E14]           1   \n",
       "Iluma-B1                                        [AAA027L17]           1   \n",
       "Luna1-A2                                        [AAA028P02]           1   \n",
       "acI-A1                               [AAA027M14, AAA278O22]           2   \n",
       "acI-A5                               [AAA028G02, AAA044O16]           2   \n",
       "acI-A6                               [AAA028E20, AAA028I14]           2   \n",
       "acI-A7         [AAA023J06, AAA024D14, AAA041L13, AAA044N04]           4   \n",
       "acI-B1    [AAA023D18, AAA027J17, AAA027L06, AAA028A23, A...           6   \n",
       "acI-B4                                          [AAA044D11]           1   \n",
       "\n",
       "         Max ANI Min ANI  \n",
       "Iluma-A2   99.97   99.97  \n",
       "Iluma-B1   99.54   99.54  \n",
       "Luna1-A2     100     100  \n",
       "acI-A1       100   79.56  \n",
       "acI-A5     99.95   86.69  \n",
       "acI-A6       100   81.17  \n",
       "acI-A7       100   87.13  \n",
       "acI-B1       100   79.06  \n",
       "acI-B4       100     100  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir, 'ANI_out', 'taxonomySAGs.csv')\n",
    "ANI.sameTribePairwiseANI(externalDataDir, pairwiseANI, taxonClass, tribes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This suggests a cutoff of 79% ANI as a reasonable cutoff for merging an additional genome into the tribe."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, let's take a look at our phylogenetic tree and identify opportunities to merge genomes. A new genome which gets added to the tribe must:\n",
    "* have ANI above the tribe-specific cutoff\n",
    "* maintain the tribe as a monophylogenetic group\n",
    "\n",
    "![Phylogenetic tree](imageFiles/2015-07-03 Phylogenetic Tree.png)\n",
    "\n",
    "First we will define a function to compute the minimum pairwise ANI within a particular tribe. The function will also compute the minimum pairwise ANI with the new genome added. If this value is lower than the cutoff defined above, then the additional genome is too divergent to be added to the tribe."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cases Involving a Tribe with Multiple SAGs\n",
    "One candidate for addition to a tribe is MEint2297 to tribe acI-A1. Let's check it out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                  AAA027M14  AAA278O22  MEintmetabat2297\n",
      "AAA027M14            100.00      79.56             75.47\n",
      "AAA278O22             79.71     100.00             75.62\n",
      "MEintmetabat2297      75.66      75.52             99.44\n",
      "79.56\n",
      "\n",
      "The minimum ANI within the tribe acI-A1 is: 79.56\n",
      "\n",
      "When genome MEintmetabat2297 is added, the min pairwise ANI becomes: 75.47\n"
     ]
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir, 'ANI_out', 'taxonomy.csv')\n",
    "ANI.addGenomeToTribe(pairwiseANI, taxonClass, tribes, 'acI-A1', 'MEintmetabat2297')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The GFM has a smaller pairwise ANI than the cutoff (79.06), so it will not get added to the acI-A1 tribe."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cases Involving a Tribe with a Single SAG\n",
    "\n",
    "For some tribes we only have a single SAG, such as AAA044-D11 for tribe acI-B4. This SAG forms a monophyletic group with the GFM MEint4252. Can we assign them to the same tribe?\n",
    "\n",
    "The function below takes two inputs:\n",
    "* existingGenomes - one or more genomes defined as belonging to the same tribe\n",
    "* newGenome - the genome being considered for addition to a cluster containing existingGenomes\n",
    "and computes the smallest pairwise ANI among all samples belonging to the same tribe (e.g., the 'cutoff' described above). The function will also compute the minimum pairwise ANI with the new genome added. If this value is lower than the first, then the additional genome is too divergent to be added to the tribe."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check out AAA044-D11 and MEint4252:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "When genome ['MEintmetabat4252'] is added, the min pairwise ANI among all samples: 84.36\n"
     ]
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir, 'ANI_out', 'taxonomy.csv')\n",
    "\n",
    "ANI.compareSamples(externalDataDir, pairwiseANI, taxonClass, tribes, ['AAA044D11'], ['MEintmetabat4252'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we can consider MEint4252 to belong to the same tribe as AAA044-D11. The phylogenetic tree suggests that TBepi.4208 is also a candidate for this tribe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "When genome ['TBepimetabat4208'] is added, the min pairwise ANI among all samples: 76.32\n"
     ]
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir, 'ANI_out', 'taxonomySAGs.csv')\n",
    "ANI.compareSamples(externalDataDir, pairwiseANI, taxonClass, tribes, ['AAA044D11', 'MEintmetabat4252'], ['TBepimetabat4208'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So that genome should not get added to the tribe.\n",
    "\n",
    "There are other examples, but I will hold off on further analysis until I incorporate coverage into my analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Constructing Tribes de novo\n",
    "\n",
    "The above approach could also be used to cluster GFMs into tribes in the absence of SAGs. This would enable us to declare multiple GFMs as belonging to the same tribe, but without knowing what specific tribe it might be (e.g., within the acIV and acV lineages which currently lack GFMs.) \n",
    "\n",
    "For example, the acV lineage contains a monophyletic group of TBepi2973 and TBhypo3180. Let's compare these two genomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "When genome ['TBhypometabat3180'] is added, the min pairwise ANI among all samples: 99.77\n"
     ]
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir,'ANI_out', 'taxonomy.csv')\n",
    "ANI.compareSamples(externalDataDir,pairwiseANI, taxonClass, tribes, ['TBepimetabat2973'], ['TBhypometabat3180'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sample TBhypo9906 is also monophyletic with the two samples above. Should we add it?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "When genome ['TBhypometabat9906'] is added, the min pairwise ANI among all samples: 71.19\n"
     ]
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir,'ANI_out', 'taxonomy.csv')\n",
    "ANI.compareSamples(externalDataDir,pairwiseANI, taxonClass, tribes, ['TBepimetabat2973','TBhypometabat3180'], ['TBhypometabat9906'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So that genome belongs to a separate tribe. (Remember, the cutoff is 79.06% ANI)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conclusions\n",
    "\n",
    "Use of ANI seems a promising way to classify genomes into tribes. To be continued..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How Many SAGs Do We Need?\n",
    "\n",
    "In order to have a robust estimate of the ANI which defines a tribe (or clade, or lineage), we need to be confident we have enough samples (e.g., so that acculumating more SAGs doesn't change the ANI cutoff). Let's see how many samples we have from each tribe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>Num Samples</th>\n",
       "      <th>Max ANI</th>\n",
       "      <th>Min ANI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Iluma-A2</th>\n",
       "      <td>[AAA027E14]</td>\n",
       "      <td>1</td>\n",
       "      <td>99.97</td>\n",
       "      <td>99.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Iluma-B1</th>\n",
       "      <td>[AAA027L17]</td>\n",
       "      <td>1</td>\n",
       "      <td>99.54</td>\n",
       "      <td>99.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Luna1-A2</th>\n",
       "      <td>[AAA028P02]</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-A1</th>\n",
       "      <td>[AAA027M14, AAA278O22]</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>79.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-A5</th>\n",
       "      <td>[AAA028G02, AAA044O16]</td>\n",
       "      <td>2</td>\n",
       "      <td>99.95</td>\n",
       "      <td>86.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-A6</th>\n",
       "      <td>[AAA028E20, AAA028I14]</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>81.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-A7</th>\n",
       "      <td>[AAA023J06, AAA024D14, AAA041L13, AAA044N04]</td>\n",
       "      <td>4</td>\n",
       "      <td>100</td>\n",
       "      <td>87.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-B1</th>\n",
       "      <td>[AAA023D18, AAA027J17, AAA027L06, AAA028A23, A...</td>\n",
       "      <td>6</td>\n",
       "      <td>100</td>\n",
       "      <td>79.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acI-B4</th>\n",
       "      <td>[AAA044D11]</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Samples Num Samples  \\\n",
       "Iluma-A2                                        [AAA027E14]           1   \n",
       "Iluma-B1                                        [AAA027L17]           1   \n",
       "Luna1-A2                                        [AAA028P02]           1   \n",
       "acI-A1                               [AAA027M14, AAA278O22]           2   \n",
       "acI-A5                               [AAA028G02, AAA044O16]           2   \n",
       "acI-A6                               [AAA028E20, AAA028I14]           2   \n",
       "acI-A7         [AAA023J06, AAA024D14, AAA041L13, AAA044N04]           4   \n",
       "acI-B1    [AAA023D18, AAA027J17, AAA027L06, AAA028A23, A...           6   \n",
       "acI-B4                                          [AAA044D11]           1   \n",
       "\n",
       "         Max ANI Min ANI  \n",
       "Iluma-A2   99.97   99.97  \n",
       "Iluma-B1   99.54   99.54  \n",
       "Luna1-A2     100     100  \n",
       "acI-A1       100   79.56  \n",
       "acI-A5     99.95   86.69  \n",
       "acI-A6       100   81.17  \n",
       "acI-A7       100   87.13  \n",
       "acI-B1       100   79.06  \n",
       "acI-B4       100     100  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir,'ANI_out', 'taxonomySAGs.csv')\n",
    "ANI.sameTribePairwiseANI(externalDataDir,pairwiseANI, taxonClass, tribes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have the most SAGs from tribe acI-B1, and the pairwise ANI between samples ranges from 79 to 99%. Let's suppose these six SAGs are representative of the entire acI-B1 sequence space. Suppose we only sequenced two SAGs. The function below will compute all possible pairs and their max and min ANI:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       Max ANI Min ANI\n",
      "(AAA023D18, AAA027J17)     100   79.88\n",
      "(AAA023D18, AAA027L06)     100   80.01\n",
      "(AAA023D18, AAA028A23)     100   80.01\n",
      "(AAA023D18, AAA278I18)     100   80.94\n",
      "(AAA023D18, AB141P03)      100   79.45\n",
      "(AAA027J17, AAA027L06)     100   81.76\n",
      "(AAA027J17, AAA028A23)     100   81.74\n",
      "(AAA027J17, AAA278I18)     100    80.1\n",
      "(AAA027J17, AB141P03)      100   80.68\n",
      "(AAA027L06, AAA028A23)     100   99.08\n",
      "(AAA027L06, AAA278I18)     100   79.44\n",
      "(AAA027L06, AB141P03)      100   96.44\n",
      "(AAA028A23, AAA278I18)     100   79.32\n",
      "(AAA028A23, AB141P03)      100   96.36\n",
      "(AAA278I18, AB141P03)      100   79.06\n",
      "Sampling tribe acI-B1 with sample size 2.\n",
      "The worst-case pairwise minimum ANI is: 99.08\n"
     ]
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir, 'ANI_out', 'taxonomySAGs.csv')\n",
    "ANI.worstCaseANI(externalDataDir, pairwiseANI, taxonClass, tribes, 'acI-B1', 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we were especially unlucky, we would have sequenced SAGs L06 and A23, and we would have predicted a tribe-specific cutoff of 99.08 ANI. What if we instead sampled three SAGs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                  Max ANI Min ANI\n",
      "(AAA023D18, AAA027J17, AAA027L06)     100   79.88\n",
      "(AAA023D18, AAA027J17, AAA028A23)     100   79.88\n",
      "(AAA023D18, AAA027J17, AAA278I18)     100   79.88\n",
      "(AAA023D18, AAA027J17, AB141P03)      100   79.45\n",
      "(AAA023D18, AAA027L06, AAA028A23)     100   80.01\n",
      "(AAA023D18, AAA027L06, AAA278I18)     100   79.44\n",
      "(AAA023D18, AAA027L06, AB141P03)      100   79.45\n",
      "(AAA023D18, AAA028A23, AAA278I18)     100   79.32\n",
      "(AAA023D18, AAA028A23, AB141P03)      100   79.45\n",
      "(AAA023D18, AAA278I18, AB141P03)      100   79.06\n",
      "(AAA027J17, AAA027L06, AAA028A23)     100   81.74\n",
      "(AAA027J17, AAA027L06, AAA278I18)     100   79.44\n",
      "(AAA027J17, AAA027L06, AB141P03)      100   80.68\n",
      "(AAA027J17, AAA028A23, AAA278I18)     100   79.32\n",
      "(AAA027J17, AAA028A23, AB141P03)      100   80.68\n",
      "(AAA027J17, AAA278I18, AB141P03)      100   79.06\n",
      "(AAA027L06, AAA028A23, AAA278I18)     100   79.32\n",
      "(AAA027L06, AAA028A23, AB141P03)      100   96.36\n",
      "(AAA027L06, AAA278I18, AB141P03)      100   79.06\n",
      "(AAA028A23, AAA278I18, AB141P03)      100   79.06\n",
      "Sampling tribe acI-B1 with sample size 3.\n",
      "The worst-case pairwise minimum ANI is: 96.36\n"
     ]
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir,'ANI_out', 'taxonomySAGs.csv')\n",
    "ANI.worstCaseANI(externalDataDir,pairwiseANI, taxonClass, tribes, 'acI-B1', 3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we were again unlucky, we would have sequenced SAGs L06, A23, and P03, and we would have predicted a tribe-specific cutoff of 96.36 ANI. Let's define a new function which repeats this calculation for all possible sample sizes (up to 6 for tribe acI-B1)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sampling tribe acI-B1 with sample size 2.\n",
      "The worst-case pairwise minimum ANI is: 99.08\n",
      "Sampling tribe acI-B1 with sample size 3.\n",
      "The worst-case pairwise minimum ANI is: 96.36\n",
      "Sampling tribe acI-B1 with sample size 4.\n",
      "The worst-case pairwise minimum ANI is: 80.68\n",
      "Sampling tribe acI-B1 with sample size 5.\n",
      "The worst-case pairwise minimum ANI is: 79.45\n",
      "Sampling tribe acI-B1 with sample size 6.\n",
      "The worst-case pairwise minimum ANI is: 79.06\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYMAAAEPCAYAAACgFqixAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFTNJREFUeJzt3Xm0HFWdwPHvey/7zlNIgiwvhKMoCIjgQYHhgYyiI7ig\nIAKC4jiIgIoKZHQgjjMMMjo6HrcZRAgzgrgggqNIWKMzY9iyQoLDEiEIJKzZSEJIzx+32u48+nW/\nel3Vt1+97+ecOl1V3V39qyz3V/feqntBkiRJkiRJkiRJkiRJkiRJyt0PgCeBJVX7uoG5wB+AG4Ep\nVe/NAv4PWA68rUUxSpJydgjwBrZNBhcD5yTr5wIXJeuvAxYCI4Ee4AGgsyVRSpJy18O2yWA5MDVZ\nn5ZsQ6gVnFv1uRuAA/MOTpIUtPrqeyqh6YjktZwYdgRWVn1uJfCqFsYlScNazKaYUrLUe1+S1AIj\nWvx7TxKah54ApgOrkv2PATtXfW6nZF9fDwAz8wxQkgroQWD3mAH08PIO5HLfwHm8vAN5FDCDEHhH\njeMVvbYwO3YAOZsdO4CczY4dQM5mxw4gR7NjB5CzhmVnnjWDq4BDgVcCjwLnEwr/HwOnAiuAY5PP\n3pfsvw/YApxO8Qt+SWobeSaD4/vZf0Q/+y9MFkmS6ipobaHrBBj7PHS8BJNvA14RO6Kc9MYOIGe9\nsQPIWW/sAHLUGzuAnBWu7CzcCQEHwOT1cE8JNpTgtM0w+abYQUkqlMKVnYU7IeBzcMYmKJXC8mwJ\nRm6MHZSkQmlYdjrkQ3yr4e4XYWuyuRgY/XzMgCSp3RWxZjAaJt4Fb1oLp2yACeuBo2IHJalQGpad\nte7lb2clhl7MAzEKeB+h4/h2YGnccCQVTOHKziLWDCQpb/YZSJIaMxlIkkwGkiSTgSQJk4EkCZOB\nJAmTgSQJk4EkCZOBJAmTgSQJk4EkCZOBJAmTgSQJk4EkCZOB8tcBY74A456DsWtg/DeArthBSRra\nnM9gyOk6GXZbB8tLsKIEb1wPY78YOyppmClc2Vm4Eyq+7utgTglKyTK3BN0LYkclDTNObqPYNq+C\n5S9Vtu8vwdbV8eKRVATWDIaeXWHc03DCC/CxTTB2LbB37KCkYaZh2TnUJkgu3KTOw8Q04DhCx/HP\ngYfjhiMNO4UrO60ZSFJ69hlIkhozGUiSoiWDTwFLgKXJOsBsYCWwIFmOjBKZJKkl9iIkgjGEDsW5\nwEzgAuDsBt+1z0CS0mtYdo5oRRR97AHMBzYm27cD70vWC9XbLUlDRYxmoqXAIUA3MA54J7Bz8t6Z\nwCLgUmBKhNgkaViKdSX+UeB0YD1wL7AJuBB4Knn/y8B04NQ+3ysBX6ravi1ZJEkVvclSdgFDoOXl\nQuC0Pvt6CP0KfdlnIEnptW3ZuUPyuguwDJhEqAmUfQa4ssb32vaEJKmNtW3ZOY/QPLQQOCzZdwWw\nmNBncC0wtcb32vaEJKmNFa7sLNwJSVILOByFJKkxk4EkyWQgSTIZSJIwGUiSMBlIkjAZSJIwGUiS\nMBlIkjAZSJIwGUiSMBlIkjAZSJIwGUiSMBlIkjAZSJIwGUiSMBlIkjAZSJIwGUiSMBlIkjAZSJJI\nlwzG5RaFJKntvQW4D3g02d4X+E6kWEqRfleShrJMys47gF2ABVX77s3iwINgMpCk9BqWnQNtJnqk\nz/aW9LFIktrViAF85hHgoGR9FHAWsCy3iCRJbWl74EpgFbAa+CHwikix2EwkSekVruws3AlJUgtk\n0mcwE7geeIpQM/gFsFtzcUmShpr5wEnAyGQ5MdnXjE8BS4ClyTpANzAX+ANwIzClxvesGUhSepmU\nnYtr7FvUxPH2IiSCMUAXIQHMBC4Gzkk+cy5wUY3vmgwkKb2mmom6CR3FvwZmAT3Jcm6yb7D2INQs\nNgIvAbcDxwBHA3OSz8wB3tPEb0iSUuio894KameTjmT/jEH+5h6Efoc3ExLCTcBdhKao7ap+45mq\n7bJSg5glSS/XsOys95xBT6ahVCwHvkLoF1gPLCTUEKqV6L9aM7tq/bZkkSRV9CbLgNXLFIcDtxCa\ncGoVzNek+aE6/hFYSehI7gWeAKYDtxJqEdWsGUhSek3VDA4lJIOjyD4Z7EB4iG0X4H3AgYRmp5MJ\ntYaTgWubOL4kKUOdwHE5HHceYbC7hcBhyb5uQv+Bt5ZKUrYyKTvvzuIgGTEZSFJ6DcvOgbS/X0R4\n+vhqQodv2TODDKoZ9hlIUnoNy856b14OnEL/t5gO9tbSZpgMJCm9psrOBY0/0nI2E0lSeg3Lznp3\nE40F9qvz/j2pw5EktaV61Ya1hCeD+3NYnffyYjORJKVnM5EkKbs5kCVJw9TbYwdQgzUDSUqvcGVn\n4U5IklrAZiJJUmONkkEX8NVWBCJJam+/p31u57SZSJLSa+qhs7KFhJnJfgJsqDpwVvMZSJIiG0gy\nGEMYlO7wPvtNBpKkKGwmkqT0MrmbaGfg58DqZPkZsFNzcUmShpqbgI8AI5PlFGBupFisGUhSepmU\nnYsGuK8VTAaSlF4mzURPAycRnjkYAZxImPlMkjSM9ADXU+kz+AWwS6RYrBlIUnpNlZ1fSV6PzSCQ\nrJgMJCm9psrOpYQnj9tpXgOTgSSl19QTyL8GngUmEGY963vgSYOPS5I01FwXO4Aq1gwkKb3ClZ2F\nOyFJagHnM5AkNWYykCTVTQY3J68XtyIQSVI89e4mmg68BTga+BHhNtPqdqd7coxLktQmPgDcQLit\n9NYaSzNmAfcCS4ArgdHAbGAl4bmGBcCRNb5nB7IkpddU2Xlw8np+BoFU6wEeIiQAgKuBk4ELgLMb\nfNdkIEnpNXU30b8mr+/NJpY/WwO8CIwjNFONAx5L3muXuZYlaVip12ewBbgEeBXwTbYtqEvAWYP8\nzWeArwGPAC8AvyHMmXAQcCbwYeAu4LPAc4P8DUlSCvVqBu8i3FH0AnB3jWWwZgKfJjQX7UgY7uIE\n4LvADGBf4HFCwpAktUC9msFqwl1Ey4GFGf7m/sD/EOZJALiGcNfSD6s+833CsNm1zK5avy1ZJEkV\nvcmSqaznQN6HMCLqWELT0xzgk8C0qs98hnCXUV92IEtSepmUnXnMgXwOlVtL5wCjgCuAxYQpNa8F\nptb4nslAktJzDmRJknMgS5Iy0oNzIEvSUJZJ2XnQAPe1gslAktLLpOysNQdyrHmRTQaSlF5TcyC/\nmXD///aEMYPKTyBPxHkQJKlQ6iWDUYSCvyt5LVsDvD/PoCRJ7WfX2AFUsZlIktJrWHYOZJTQWnMX\nlIDDU4fTvBKObCpJaTUsO+s1E5V9vmp9DHAMYURTSdIwd2ek37WZSJLSa+puorLuqvVOwqijkwYb\nkSSp/QwkGdxDJatsAVYAp+YVkCRJjdhMJEnpNTVQ3UmEKShr7f/QYCOSJA0td7Dtw2ZlEwhNRzFY\nM5Ck9JqqGYwE1tbYvy55T5JUEPWSwRhCLaCviZgMJKlQ6iWDS4GfEOYzKJsBXJ28J0kaJk4D/gg8\nkyyPAJ+IGI99BpKUXmZl5yTa40Ezk4EkpVe4srNwJyRJLdDU3USSpGHCZCBJqjs20TH0PwZ2Cbgm\nl4gkSS1XLxkcRf12JpOBJCkKO5AlKb2m5jM4ucYBOqr2XTHIoKSieQdM+hBsWQMbvgY8FDsgKa16\nc2J+i9rJ4ChgJ6Arr6DqcA5ktZmOD0H3JfClcfCnrfD1tfDCvoR5P6R2kVnZ2QmcCCwhDEexdxYH\nHQSbidRmpjwEt5SglCyf2gJd/xA7KqmPpqe9HEloLvocMB94P3B/83FJRVEaBVOqtrfrhK4x8FK0\niKSsnQH8AfguYYC6LM0C7iXUNK4ERhPmWp6b/OaNbPs/rMyagdrMuNmw53qYV4KrSzBhPWGecKmd\nNCw767UhbQVWAav7OfBgm4p6gFuA1wKbCM1OvwL2BJ4CLgbOBbYDzqvxu/YZqJ10wthzYeyHgXXw\nzHnAzbGDkvpoquzsqbPs2kRQ3YSmpu0IzVTXA38JLAemJp+Zlmz3Zc1AktLLpew8BPh2k8f4OGEW\ntVXAfyT7nq16v6PPdpnJQJLSa7oDuWw/4HjgWOBh4GdNBDUT+DShhvE8YQKdE/t8pkT/wc+uWr8t\nWSRJFb3JkonXEAreZcA84EzC5DbNOg74ftX2SYSaxjJC8xDAdGwmktpFJ+H/5OjYgWjQmio7twLX\nAbtU7Xu4qXCCfYClwFhCc9Ac4JNUOo4hdBxfVOO7JgOptfaC8X+CiRtg1EYYcUrsgDQoTZWd7yHc\n6bMC+B7wVrJ7qvIcKreWziE8z9AN3IS3lkrtogPGr4RLt4YH6paVQlJgz9iBKbVMys4JwAnAL4H1\nhOcO3pbFgQfBZCC1zmQYtbnydHWpBEevIZQHRdIFbE+x53fJvOzsJtwJdEvWBx4gk4HUOp0weh3M\nTxLBmhJMXwccHDuwDB0Jo9fCuI0w9mngwNgB5aRwZWfhTkhqc0fDuPVw2HOwwzqY+G+xA8rQVBi3\nDn6XJLtrSzD2WWBM7MByULiys3AnJA0BPYQ+xAMix5G1w2G/57ZtBpu2lnAnZdEUruws3AlJiubV\nMGkDrEoSwYMlGL2R2jevDHWFKzsLd0KSYhr/ZeheD29/Hiath9GfiB1RTpoaqK4dOVCdpKztRxgZ\n4V7gvsix5KVwZac1A0lKr2HZWeT7aiVJA2QykCSZDCRJJgNJEiYDSRImA0kSJgNJEiYDSRImA0kS\nJgNJEiYDSRImA0kSJgNJEiYDSRImA0kSJgNJEiYDSRImA0kSJgNJEiYDSRImA0kSJgNJEiYDSRIw\nIsJvvgb4UdX2bsD5wHbAx4DVyf5ZwA2tDU2SFEMn8DiwM3ABcHaDz5dyj0iSiqdh2Rm7megI4AHg\nUaAjWSRJLRY7GXwQuCpZLwFnAouAS4EpsYKSpOEm5pX4KOAx4HWEfoIdqPQXfBmYDpza5zsl4EtV\n27cliySpojdZyi6gjVte3k3/HcQ9wJIa++0zkKT02rrP4HgqTUQQagJl76V2MpAkFch44ClgYtW+\nK4DFhD6Da4GpNb5nzUCS0itc2Vm4E5KkFmjrZiJJUpswGUiSTAaSJJOBJAmTgSQJk4EkCZOBJAmT\ngSQJk4EkCZOBJAmTgSQJk4EkCZOBJAmTgSQJk4EkCZOBJAmTgSQJk4EkCZOBJAmTgSQJk4EkCZNB\nu+mNHUDOemMHkLPe2AHkrDd2ADnqjR1AbCaD9tIbO4Cc9cYOIGe9sQPIWW/sAHLUGzuA2EwGkiST\ngSQJOmIHkNJCYJ/YQUjSELMI2Dd2EJIkSZIkSVJGfgA8CSyJHUhOdgZuBe4FlgJnxQ0nc2OA+YQ+\nn/uAf4obTi66gAXA9bEDycEKYDHh/O6IG0oupgA/BZYR/n0eGDecTL2G8PdWXp5niJcvhwBvoLjJ\nYBqVzp0JwP3Aa+OFk4txyesI4PfAwRFjycPZwA+B62IHkoOHge7YQeRoDvDRZH0EMDliLHnqBB4n\nXHzWfHMo+C3wbOwgcvQE4aoZYB3hCmXHeOHkYkPyOopwFf1MxFiythPwTuD7DL079AaqqOc1mXCx\n+YNkewvh6rmIjgAeBB6t9eZQSQbDSQ+hFjQ/chxZ6yQkvCcJTWL3xQ0nU18HPg9sjR1ITkrATcBd\nwF9HjiVrM4DVwGXAPcAlVGqxRfNB4MrYQWShh+I2E5VNIPyHe0/sQHI0mdBM1Bs5jqy8C/h2st5L\nMfsMpiev2xMS+iERY8na/sCLwAHJ9jeAv48XTm5GEZLe9v19wJpB+xgJ/Az4T+DayLHk6Xngvwj/\nCYvgLcDRhHb1q4DDgSuiRpS9x5PX1cDPgTdFjCVrK5PlzmT7p8B+8cLJzTuAuwl/h0NeD8WtGXQQ\nCpCvxw4kJ68k3LEBMBaYB7w1Xji5OZTi1QzGAROT9fHAfwNvixdOLuYBr07WZwNfiRdKbn4EnBw7\niCxcBfwJ2ETo/PhI3HAydzChvXkhlVvAjowaUbZeT2iPXUi4RfHzccPJzaEU726iGYS/t4WE255n\nxQ0nF/sQagaLgGso3t1E44GnqCR1SZIkSZIkSZIkSZIkSZIkSRoqtgJfrdr+HHBBRse+HDgmo2PV\n8wHCuEs399nfCXyT8ODkYsJQ0D05x7KCYo82qgw5HIXayWbgvcArku1Shsdu5lgjUnz2VOBjvPwJ\n6+MIY/y8HtibMP7Uc03ENBBZ/vmp4EwGaicvAv8OfKbGe5ez7ZX9uuS1F7idMJ7Tg8BFwEmEK+/F\nwG5V3zmC8KTp/cBfJfu6gH9OPr8I+HjVcX8L/IIw6VBfxyfHX5L8JsD5wEGE4ZAv7vP5aVTG+IHw\nRH05GXwniWspYTiEshXAhYQn0u8ijJlzI/AA8DdVcc4DfgksB75L7eGmTySMhLsA+B7h/34X4c+1\nXFv5dI3vSVLLrSU8Mv8wMAn4LJVmosvYNhmsTV57CXNdTCWMzPgYlQL1LCrjPV0O/CpZ350wrMlo\nQuH/hWT/aEKh3JMcdx2wa404dwT+SKjBdBGahN6dvHcrtQc6e1VyXgsITWH7Vr23XfLalXx/r2T7\nYSqF/r8QCuzxhLGenqg6/xeSmDsJyeKYqu93EyZKui45PoRRVk+iklzKijYMg1KwZqB2s5YwaF+a\nqfnuJMyTsJlw1fybZP9SKu3yJeDHyfoDwEPAHoRB1z5MKKR/Tyg8d08+dweh0O/rAEKh/TTwEmGG\ns7+oer/WlfljhCkIZxH6Rm4mjHAKoQnpbsL4TXsCr6v6XnmsoyXA/wLrCePMbCIkzHKcK5LjXsW2\ns8h1EJqs3kioXSxItmckfwa7Efoy3g6sqRG3hok0baFSq3yDUDBeVrVvC5WLl05CLaBsU9X61qrt\nrdT/N15uUz8DmNvnvV5Cwdvf96oL/A62bZ/vr61+M3BDsjxJ6Dd4iFAD2p8wvPdlhDmjy6rPZXPV\n/upzq/69DmpPsjMH+Nsa+/cmDIp4GnAsoc9Dw5A1A7WjZwlX8adSKehWEK5uIcwfMDLlMTsId/p0\nADMJV8TLCbWI06kUrK+m8UxXdxJGKC03E32Q0G9RzxuoTGXaSRgpcwXh6n494ap8KmHc+f7i78+b\nqDQTHQf8ruq9EqEW8n4qE5t0A7sk8Y8gjNT5dxRzHH8NkDUDtZPqK9yvEa7Yyy4hdOYuJFxZr+vn\ne32PV6paf4TQpDKJ0Ba/mTBvcQ+hJtIBrCLc0VT93b4eB84jNBV1EDpvG81jsENyDqOT7fnAt5IY\nFhAS06NsW5D3dy70Wb8zOdbuwC2ECWiqP7MM+CKhf6CT0FF/OrCRUBMpXxSe1+AcJEltqpfiTaij\nCGwmkoa2ejUYSZIkSZIkSZIkSZIkSZIkSUrr/wHOYCqSJXW9HwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1066b6a50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir,'ANI_out', 'taxonomySAGs.csv')\n",
    "ANI.sameTribePairwiseANI(externalDataDir,pairwiseANI, taxonClass, tribes)\n",
    "ANI.allWorstCaseANI(externalDataDir, pairwiseANI, taxonClass, tribes, 'acI-B1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These six samples appear to be enough. What about tribe acI-A7, for which we have 4 samples?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sampling tribe acI-A7 with sample size 2.\n",
      "The worst-case pairwise minimum ANI is: 96.2\n",
      "Sampling tribe acI-A7 with sample size 3.\n",
      "The worst-case pairwise minimum ANI is: 88.21\n",
      "Sampling tribe acI-A7 with sample size 4.\n",
      "The worst-case pairwise minimum ANI is: 87.13\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEPCAYAAABY9lNGAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFSBJREFUeJzt3Xu4HGV9wPHvyYWQkGAISBIJeCAoXrlYpQpSF1S03qoF\nRdsgWqwiKl5agdQLx14oBas+1lsrCmgREUVuVQuCUavlJgESQgQ0Ry5KAnILBwVDtn+873Y3m3f3\n7O6ZOTN79vt5nn3O7szszm9ewvub952Z9wVJkiRJkiRJkiRJkiRJkiRpoHwZWA+sali2ALgMuAW4\nFJjfsG45cCuwFjh0kmKUJBXgIGA/tkwQpwLHx/cnAKfE988ArgdmAsPAbcC0SYlSklSIYbZMEGuB\nhfH9ovgZQuvhhIbtvgc8P+/gJEmtTfZZ+kJCtxPxby1ZPAm4s2G7O4FdJjEuSVKTIrtxqvHVbr0k\nqSAzJnl/6wldS3cDi4ENcfldwK4N2y2Jy5rdBizNM0BJmoJ+AexZdBDNhtn6InXtWsOJbH2Rehtg\nd8LBDCV+r19aFSNFB9ChkaID6NBI0QF0aKToADo0UnQAHRopOoAOjBQdQId6qjvzbEGcA7wI2Am4\nA/goISF8AzgaGAXeELddE5evATYBx9I/yUCSpqQ8E8SbWix/SYvlJ8eXJKkEfNYgHyuKDqBDK4oO\noEMrig6gQyuKDqBDK4oOoEMrig6gAyuKDkB1djtJUvd6qjttQUiSkkwQkqQkE4QkKckEIUlKMkFI\nkpJMEJKkJBOEJCnJBCFJSjJBSJKSTBCSpCQThCQpyQQhSUoyQUiSkkwQkqQkE4QkKckEIUlKMkFI\nkpJMEJKkJBOEJCnJBCFJSjJBSJKSTBCSpCQThCQpyQQhSUoyQUiSkkwQkqQkE4QkKckEIUlKMkFI\nkpJMEJKkJBPEYHop7LASdrgFZn8E/x1ImgKqRQcwBewPc8fgm1X4SRWePQZz/r7ooCTlaiDqzoE4\nyHzN+gR8bDNUq+F1XRWecEfRUUnKVU91p10LA+fxR+C+zfXP9wNDjxYWjiRlxBbExD0ZZj8AH9wE\nn6rCgjEYemPRQUnK1UDUnQNxkJNgd9j2UzDvDODQooORlLuBqDsH4iAlKWNeg5AkZccEIUlKMkFI\nkpJMEJKkpKISxHuBVcDq+B5gBLgTWBlfLy8kMklSYZ5FSA7bAtOBy4ClwEnAB8b5rncxSVL3eqo7\nZ2QdRQeeBlwF/D5+/iHw5/H9UAHxSJISiuhiWg0cBCwA5gCvAHaN694D3AB8CZhfQGySpKioM/a/\nAo4FxoCbgEeBk4F74/p/ABYDRzd9rwp8rOHziviSJNVV4qvmJPq0h+Zk4JimZcOE6xTNvAYhSd3r\nq7pz5/h3N+BmYHtCi6Hm/cDXEt/rq4OUpJLoq7rzR4SupeuBg+OyrwA3Eq5BXAAsTHyvrw5Skkpi\nIOrOgThIScqYg/VJkrJjgpAkJZkgJElJJghJUpIJQpKUZIKQJCWZICRJSSYISVKSCUKSlNRNgpiT\nWxSSpL50ALAGuCN+3hf4XEGxONSGJHUvt7rzasKoqysblt2U187GYYKQpO7lOhbT7U2fN/WyM0lS\n/+hkTurbgQPj+22A4whzOEiSBtwTCZP3bADuAc4GdiwoFruYJKl7A1F3DsRBSlLGcrsGsRS4GLiX\n0IK4ENijl51JkqaWq4AjgZnxtSwuK4ItCEnqXm51542JZTfktbNxmCAkqXs91Z3t7mJaAAwB3wWW\nA+fE5UfEZZKkKWyozbpR0llnKC7fPY+AxlGlfcySpK0NRN1pF5MkdS/zLqZDgCuAw1r8+Pm97FCS\n1B/aJYgXERLEqzFBSJKaTCNclC4Lu5gkqXu51Z0/y+uHe2CCkKTu9VR3dnJV+xTCU9TnAmMNy+/r\nZYcTNBBX4iUpYz3Vne2+cCbwFlrf7uptrpLUHzKvO1eOv8mks4tJkrqX+W2us4HntFl/XS87lCT1\nh3ZNjo3AtW3WH5xxLJ2wi0mSumcXkyQpKdc5qSVJ+n8vKzqABFsQktS9gag7B+IgJSljdjFJkrIz\nXoKYDnx8MgKRJPWfKynPraV2MUlS9zJ/UK7meuBC4DzgkYadOdy3JE1hnSSIbQkD8x3StNwEIUkq\nDbuYJKl7ud3FtCvwbeCe+PoWsKSXnUmSppbvA28FZsbXW4DLCorFFoQkdS+3uvOGDpd1473AKmB1\nfA+wgJB4bgEuBeYnvmeCkKTu5dbF9FvgSMIzETOAZYQZ5nr1LOBtwPOAfYBXAUuBEwkJ4qnA5fGz\nJKnEhoGLqV+DuBDYbQK/dzhwesPnDwPHA2uBhXHZovi5mS0ISepe5nXnv8S/b8j4d58G/JzQpTQH\n+CnwaeD+hm2Gmj7XmCAkqXuZPyj3SkI3z3LgG738eAtrCcnnUmCM8CDe403bVGl9QCMN71fElySp\nrhJfuTkNeADYRJhdrvH1UIb7+SfgnYTEsSguW4xdTJKUldzqzoty+M2d49/dgJuBJwCnAifE5ScC\npyS+Z4KQpO71Vd35I+AmQvdSbW7rBYRnLrzNVZKyNRB150AcpCRlzAmDJEnZaZcgLo9/T52MQCRJ\n5dLuNtfFwAHAa4CvE55NaGymXJdjXJKkgrWbKe71wNHAgcC1ifUHJ5blrUp5ZreTpH6Red35wvj3\no1n+6AR5kVqSupd53fmz+Hdl1j88ASYISepe5kNtbAK+COxCGCupsXlSBY7rZYeSpP7QLkG8Cngx\ncCihNdGcICRJA27fogNoYGKSpO7lOmGQc1JLkrbinNRSezOBZxIm15LKqK/mpO6VCUJlswvM/QUs\n3gjzfgfzvo5D2Kh8cqs7r2DrOakvb/uN/JggVDLzL4MP/QE2V2GsCvs9THjAVCqT3OrOYbKdk3oi\nTBAqmXl3w9oqVOPrtCrM/kzRUUlNcqs7D+xw2WQwQahk5v8QTt4UksPvq/CCMeCYoqOSmuRWd6ae\npC7q6WoThMpmd9juLnjKg7DTGGx/Ce2fL5KKkPmT1C8gjOb6ROAD1B+Um4cX4aSadTD2FLh1b2AM\nWI0nMpoi2iWIbQjJYHr8W/MQcHieQUl95hHgyqKDkIrw5KIDaOCZmSR1r6e6s5PxwX/QYmeH9LLD\nCXI+CEnqXk91ZycX0z7Y8H5b4DDCSK+SJG3lmoL2axeTNBi8EyxbuQ3Wt6DhtRPwcmD7XnYmSeN4\nNswdhaHHYLv11Ge2VEmNAuvi61bCQH1F/UezBSFNXbNgzgY4YzM8XoVLqjD7IcKJqSZmIOrOgThI\naUDtBYs21octqVZh3weAg4sObArIvIvpSODNLZb/RS87k6Q27oX7Z8Kv48cHgdFtgPUFxqQWrmbL\nB+Rq5gLXTXIsNbYgpCltznLY6WFYNgZLHoZ5ny06oiki87qz3XhLq7LeWYdMENLU9wLgHRTzrNVU\nlXndeTOhtdBsHrA26511yAQhSd3L/BrEl4Dz2HIaxd2Bc+M6SdIAOwb4FXBffN0OvLPAeGxBSFL3\ncq07t6ccD8eZICSpewNRdw7EQUpSxnIbakOSNIBMEJKkpHYjJh5G6zHEq8D5uUQkSSqFdgni1bTv\ntzJBSJJKw4vUktS9nurOdi2IoxI/OtSw7Cu97FCS1B/azVH6GdIJ4tXAEmB6XkG14ZzUktS9XOvO\nacAywiB95wJ757WjcdjFJEndy6XunAm8jTA431nAXnnspAsmCEnqXuZ157uBW4DPEwbpy9Jy4CZC\ni+RrwCxgBLiTMMz4SsLc181MEJLUvZ7qznZ9UpuBDcA9LXbWazfTMHAF8HTgUUKX1Xfi8o3AJ9p8\n12sQktS9nurOdncx7THOznr1EPAHYA7wePx7FyFBWPlLUh87CJjoNIBvJ7QWNgBfjctOAkaBGwjz\nTcxPfM8uJknqXq5153OA0whzQ6wA3jOB31oKrAF2JLRgvg38JbAzoQUxBPwj6UmJTBCS1L3MH5Tb\nC3gTcAThOsR5hMq70suOGjwX+Cnw2/j5fOAA4OyGbU4HLm7x/ZGG9yviS5JUV2HidXVbm4GLgN0a\nlq3L4Hf3AVYDswkJ5yzgXcCihm3eT7i7qZktCEnqXuZ152sJdxiNAl8AXhzfZ+F46re5ngVsQxi6\n40bCNYgLgIWJ75kgJKl7udWdcwnXCC4BxgjPRRya187GYYKQpO5NSt25gHAH0hWTsbMEE4QkdW8g\n6s6BOEhJyphzUkuSsmOCkCQlmSAkSUkmCElSkglCkpRkgpAkJZkgJElJJghJUpIJQpKUZIKQJCWZ\nICRJSSYISVKSCUKSlGSCkCQlmSAkSUkmCElSkglCkpRkgpAkJZkgJElJJghJUpIJQpKUZIKQJCWZ\nICRJSSYISVKSCUKSlGSCkCQlmSAkSUkmCElSkglCkpRkgpAkJZkgJElJJghJUpIJQpKUZIKQJCWZ\nICRJSSYISVKSCUKSlGSCkCQlmSAkSUkmCElSkglCkpRUVIJYDtwErAK+BswCFgCXAbcAlwLzC4pN\nklSQYeCXhKQAcC5wFHAqcHxcdgJwSuK71byDk6QpqG/qzgXAz4EdgBnAxcBLgbXAwrjNovi5Wd8c\npCSVSF/VnW8HNgIbgK/GZfc3rB9q+lzTVwcpSSXRU905I+soOrAUeB+hq+lB4DxgWdM2VVof0EjD\n+xXxJUmqq8RX3zkCOL3h85HAZ4GbCV1LAIuxi0mSstJT3VnEXUxrgecDswldSS8B1hCuRRwVtzkK\nuKCA2CRJBTue+m2uZwEzCRevv0/721xtQUhS9wai7hyIg5SkjPVNF5MkqQ+YICRJSSYISVKSCUKS\nlGSCkCQlmSAkSUkmCElSkglCkpRkgshHpegAOlQpOoAOVYoOoEOVogPoUKXoADpUKTqADlSKDiBP\nJoh8VIoOoEOVogPoUKXoADpUKTqADlWKDqBDlaID6ECl6ADyZIKQJCWZICRJSUNFB9Cl64F9ig5C\nkvrMDcC+RQchSZIkSZIklcCXgfWE2eZa+TRwK6Ffbb/JCCphvDgrwIPAyvj68OSEtZVdgR8QZvBb\nDRzXYruiy7STOCsUX6bbAlcRroetAf65xXZFl2cncVYovjwBpsf9X9xifdFlWdMuzgrlKMtR4MYY\nw9UttilLefbkIELQrSreVwDfie//GLhyMoJKGC/OCnDRpEXT2iLqF6fmAj8Hnt60TRnKtJM4K5Sj\nTOfEvzMIZfXCpvVlKE8YP84K5SjPDwBnk46lLGUJ7eOstFg+2dYRpm9upavyLONtrj8G7m+z/jWE\neawhnCHNBxbmHVTCeHFCOe4Su5twFgnwMHAz8KSmbcpQpp3ECeUo00fi320IZ5X3Na0vQ3nC+HFC\n8eW5hFBpnd4ilrKU5Xhx0mb5ZGsXR1flWcYEMZ5dgDsaPt9J+I9XNlXgAEIz7jvAM4oNB4BhQqvn\nqqblZSvTYdJxlqVMpxGS2XpCt9iapvVlKc/x4ixDeX4S+CCwucX6spTleHGWoSxrcXwfuBb468T6\nrsqzHxMEbJ0he5qQO2fXEfrV9wH+Dbig2HCYC3wTeC/hDL1ZWcq0XZxlKdPNhO6wJcCfkB5uoQzl\nOV6cRZfnq4ANhP7ydme9RZdlJ3EWXZY1BxJOrv4UeBehK7xZx+XZjwniLsJ/iJolcVnZbKTexP8u\nMJP2fYN5mgl8C/hP0v9wy1Km48VZpjKFcFHyv4DnNi0vS3nWtIqz6PI8gNDlsQ44BzgE+ErTNmUo\ny07iLLosa34T/94DfBvYv2l9Gcpzwobp7CL18yn2otUwreNcSD1T70+4u6AIQ4R/zJ9ss00ZyrST\nOMtQpjsR+m0BZgM/Al7ctE0ZyrOTOMtQnjUvIn13UBnKslGrOMtQlnOAefH9dsBPgEObtumqPGdk\nGV1GziH8R9iJ0Fd2EiEbA/w74eBeAdwGjAFvLSBGGD/Ow4F3ApsIZxZvLCBGCE3OZdRvfQP4O2C3\n+L4sZdpJnGUo08WEi3zT4uurwOXAO+L6spRnJ3GWoTwb1bo6ylaWzVJxlqEsFxJaDRDq9rOBSyl/\neUqSJEmSJEmSJEmSJEmSJElSv9oMfLzh898Sni/JwpnAYRn9VjuvJ4x5dHnT8mmEYZZXEZ71uJrw\noGWeRin2SXP1uX4cakNT12PA64Ad4+csx9yZyG9180Dp0cDb2Pqp5SMID689G9gbeC3wwARi6kQZ\nxyhTHzFBqEz+APwH8P7EujPZsgVQG8ivAvyQMHbTL4BTgCMJZ+g3Ans0fOclwDWEuSZeGZdNB06L\n298AvL3hd38MXEiYxKjZm+Lvr4r7BPgo4YnwLwOnNm2/iPo4OQC/pp4gPhfjWg2MNGwzCpxMeLL8\nWuA5hCdjb6P+dGyFMIzGJcBa4POkB5RbRhgddyXwBcL/+9MJ5Vpr1bwv8T1JKoWNhLFk1gHbA39D\nvYvpDLZMEBvj3wphXo6FhHkP7qJeyR5HfWynM6mPQbMnYXiUWYSE8KG4fBahoh6Ov/sw8OREnE8C\nfkVo6UwndCf9WVz3A0JF3myXeFwrCd1o+zas2yH+nR6//6z4eR31RPAJQiW+HWF4l7sbjv93MeZp\nhARyWMP3FxAmXroo/j7AZwlJtJZwap6QiFsDzBaEymYjYdC+VlOjplxDmPPgMcLZ9X/H5aup9/NX\ngW/E97cBvwSeRhjM7M2EivtKQoW6Z9zuakIiaPY8QkX+W+Bxwpg3f9KwPnUGfxewF7CccK3lcsKo\noBC6n35GGDL6mWw5l0BtlrJVwP8Sxs+5F3iUkERrcY7G3z2HLWeOGyJ0d/0RoRWyMn7ePZbBHoRr\nIy8DHkrErQFWxsH6pE8RKsszGpZton5CM43QWqh5tOH95obPm2n/b7zWR/9u4LKmdRVCZdzqe41J\nYIgt+/tb9f0/BnwvvtYTrkP8ktBSei5hWO4zCPNJ1zQey2MNyxuPrXF/Q6QntTmLMPhhs72BlwPH\nAG8gXEORAFsQKqf7CWf7R1Ov/EYJZ8EQxuafufXX2hoi3GE0BCwlnDmvJbQ2jqVe2T6V+lzOrVxD\nGMm31sX0RsJ1kHb2oz6F6jTCxDKjhFbAGOHsfSFhopdW8beyP/UupiOA/2lYVyW0Vg4HnhiXLSCM\nkrsj4bjPBz5CumtMA8wWhMqk8Uz4Xwln9jVfJFwwvp5wBv5wi+81/1614f3thO6Y7Ql9+48R5hge\nJrRYhggzh72u6bvNfgOcSOhmGiJcIE7NEdBo53gMs+Lnq4DPxBhWEpLVHWxZubc6FpreXxN/a0/g\nCupDPte2uRn4MOF6wzTCzQDHAr8ntFhqJ4onjnMMkqQ+UmH85CT1xC4mqb+1a+lIkiRJkiRJkiRJ\nkiRJkiRJ0lTzf7bUdfT8+FA3AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1096aadd0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "[pairwiseANI, taxonClass, tribes] = ANI.importANIandTaxonomy(externalDataDir,'ANI_out', 'taxonomySAGs.csv')\n",
    "ANI.sameTribePairwiseANI(externalDataDir,pairwiseANI, taxonClass, tribes)\n",
    "ANI.allWorstCaseANI(externalDataDir,pairwiseANI, taxonClass, tribes, 'acI-A7')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having one more SAG would give us more confidence in the ANI cutoff."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reverse Ecology Analysis on Merged Genomes\n",
    "\n",
    "When the above analysis is complete, we will have sets of genomes whose metabolic network models can be merged to give a more comprehensive model for each tribe. For now, we will only merge graphs for those SAGs belonging to the same tribe.\n",
    "\n",
    "First import the necessary preliminaries and custom functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Import special features for iPython\n",
    "import sys\n",
    "sys.path.append('../Python')\n",
    "import matplotlib\n",
    "%matplotlib inline\n",
    "\n",
    "# Import Python modules \n",
    "# These custom-written modules should have been included with the package\n",
    "# distribution. \n",
    "import graphFunctions as gf\n",
    "import metadataFunctions as mf\n",
    "import sbmlFunctions as sf\n",
    "import seedFunctions as ef\n",
    "\n",
    "# Define local folder structure for data input and processing.\n",
    "rawModelDir = 'ProcessedModelFiles'\n",
    "processedDataDir = 'MergedData'\n",
    "summaryStatsDir = 'DataSummaries/MergedData'\n",
    "externalDataDir = 'ExternalData'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First read in a file listing {lineage, clade, tribe} for each sample and aggregate all genomes belonging to the same tribe. The current input file taxonomySAGs.csv contains classifications for the SAGs only."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing taxonomy\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Iluma-A2': ['AAA027E14'],\n",
       " 'Iluma-B1': ['AAA027L17'],\n",
       " 'Luna1-A2': ['AAA028P02'],\n",
       " 'acI-A1': ['AAA027M14', 'AAA278O22'],\n",
       " 'acI-A5': ['AAA028G02', 'AAA044O16'],\n",
       " 'acI-A6': ['AAA028E20', 'AAA028I14'],\n",
       " 'acI-A7': ['AAA023J06', 'AAA024D14', 'AAA041L13', 'AAA044N04'],\n",
       " 'acI-B1': ['AAA023D18',\n",
       "  'AAA027J17',\n",
       "  'AAA027L06',\n",
       "  'AAA028A23',\n",
       "  'AAA278I18',\n",
       "  'AB141P03'],\n",
       " 'acI-B4': ['AAA044D11']}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tribeSampleDict =  mf.importTaxonomy('../ExternalData/taxonomySAGs.csv')\n",
    "tribeSampleDict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, merge the metabolic network graphs for the samples from each tribe. Each sample is converted to a graph object and merged with the previous graph. The final graph is written to file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging genomes from individual tribes\n"
     ]
    }
   ],
   "source": [
    "gf.createTribalGraph(tribeSampleDict, processedDataDir, rawModelDir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The remaining code operates similarly to the RE analysis using all individual genomes. The metabolic network graphs corresponding to each tribe are imported:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dirList = mf.getDirList('../'+processedDataDir)\n",
    "numSubDir = len(dirList)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compute summary statistics on the size of each tribe's network graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "IOError",
     "evalue": "[Errno 2] No such file or directory: '../DataSummaries/MergedData/GraphStatistics.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIOError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-0f114f2240fa>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mgraphStatArray\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdiGraphStatArray\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcomputeGraphStats\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdirList\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprocessedDataDir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msummaryStatsDir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mgf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplotGraphStats\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgraphStatArray\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/joshamilton/Documents/Research/2014aii-ActinosRecon/2015-04-13-REMasterDirectory/Python/graphFunctions.pyc\u001b[0m in \u001b[0;36mcomputeGraphStats\u001b[0;34m(dirList, processedDataDir, summaryStatsDir)\u001b[0m\n\u001b[1;32m    308\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    309\u001b[0m \u001b[0;31m# Create files to record the summary statistics.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 310\u001b[0;31m     \u001b[0mgraphFile\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0msummaryStatsDir\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'/'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'GraphStatistics.txt'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'w'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    311\u001b[0m     \u001b[0mgraphFile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Model,Nodes,Edges,Total Components,Size of Largest\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIOError\u001b[0m: [Errno 2] No such file or directory: '../DataSummaries/MergedData/GraphStatistics.txt'"
     ]
    }
   ],
   "source": [
    "graphStatArray, diGraphStatArray = gf.computeGraphStats(dirList, processedDataDir, summaryStatsDir)\n",
    "gf.plotGraphStats(graphStatArray)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert each network graph to its largest component."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "reducedGraphStatArray = gf.reduceToLargeComponent(dirList, processedDataDir, summaryStatsDir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compute seed sets and plot summary statistics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "seedSetList = gf.computeSeedSets(dirList, externalDataDir, processedDataDir)\n",
    "gf.plotSeedStatsForTribes(seedSetList, reducedGraphStatArray)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aggregate the seed compounds for all tribes into a single matrix and write the results to file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "seedMatrixDF = ef.consolidateSeeds(dirList, externalDataDir, processedDataDir, summaryStatsDir)\n",
    "ef.normalizedSeedCounts(dirList, processedDataDir, summaryStatsDir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cluster the seed profiles and create the dendrogram:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ef.clusterSeedSets(seedMatrixDF, dirList, externalDataDir, summaryStatsDir, 'tribalColors.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genomes in the above clustergram are colored as follows:\n",
    "\n",
    "acI-A: green  \n",
    "acI-B: blue  \n",
    "acIV-A: red  \n",
    "acIV-B: maroon  \n",
    "acSTL-A: black  \n",
    "acTH1-A: grey  \n",
    "Luna1-A:  cyan  \n",
    "\n",
    "And here is just the dendrogram for genomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ef.clusterOnly(seedMatrixDF, dirList, externalDataDir, 'tribalColors.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compute the competidion index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "metabCompeteDF = ef.computeMetabCompete(dirList, processedDataDir, summaryStatsDir)\n",
    "ef.clusterPairwise(metabCompeteDF, dirList, externalDataDir, summaryStatsDir, 'tribalColors.csv', 'metabolicCompetition.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compute the cooperation index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "metabComplementDF = ef.computeMetabComplement(dirList, processedDataDir, summaryStatsDir)\n",
    "ef.clusterPairwise(metabComplementDF, dirList, externalDataDir, summaryStatsDir, 'tribalColors.csv', 'metabolicComplementarity.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are currently analyzing the seed profiles to look for shared features of the actinobacterial lineage, and unique features of each tribe. \n",
    "\n",
    "\n",
    "** TO DO **\n",
    "1. Provide an illustration of what the \"merging\" process looks like for metabolic network graphs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
